js/web/lib/wasm/jsep/webgpu/ops/attention.ts

```diff
@@ -1004,21 +1004,188 @@ const prepare = (context: ComputeContext, parameters: AttentionParameters) => {
+  const bC = bR;
+  const tR = q.dims[2] / bR;
+  const tC = k.dims[2] / bC;
+  const d = q.dims[3] / components;
+  const numTiles = Math.ceil(q.dims[3] / bC);
+  const workgroupSize: [number, number, number] = [8, 32, 1];
+  const qInner = numTiles * workgroupSize[0];
+  const colsPerThread = 4; // (Bc / workgroupSize[0])
+  const alpha = attributes.scale === 0 ? 1.0 / Math.sqrt(parameters.headSize) : attributes.scale;
+
+  const dispatch = {
+    x: tR,
+    y: v.dims[1],
+    z: v.dims[0],
+  };
+
+  const headOffset = v.dims[2] * v.dims[3];
+  const batchOffset = v.dims[1] * headOffset;
+  const programUniforms: ProgramUniform[] = [
+    { type: DataType.uint32, data: batchOffset },
+    { type: DataType.uint32, data: headOffset },
+    { type: DataType.uint32, data: v.dims[1] },
+    { type: DataType.uint32, data: v.dims[3] },
+    { type: DataType.uint32, data: d },
+    { type: DataType.float, data: alpha },
+  ];
+
+  const outputDims = [v.dims[0], v.dims[2], v.dims[1] * v.dims[3]];
+  const outputs = [{ dims: outputDims, dataType: q.dataType, gpuDataType: GpuDataType.default }];
+  const inputDependencies: ProgramInputTensorInfoDependency[] = ['type', 'type', 'type'];
+
+  const getShaderSource = (shaderHelper: ShaderHelper) => {
+    const qInput = inputVariable('Q', q.dataType, q.dims, components);
+    const kInput = inputVariable('K', k.dataType, k.dims, components);
+    const vInput = inputVariable('V', k.dataType, k.dims, components);
+    const inputVars = [qInput, kInput, vInput];
+
+    const output = outputVariable('output', q.dataType, outputDims);
+    const outputVars = [output];
+    const type = tensorTypeToWsglValueType(v.dataType);
+
+    const uniforms: UniformsArrayType = [
+      { name: 'batchOffset', type: 'u32' },
+      { name: 'headOffset', type: 'u32' },
+      { name: 'headNum', type: 'u32' },
+      { name: 'headSize', type: 'u32' },
+      { name: 'd', type: 'u32' },
+      { name: 'alpha', type: 'f32' as UniformDataElementType },
+    ];
+
+    return `
+    var<workgroup> Q_i : array<array<${qInput.type.storage}, ${qInner}>, ${bR}>;
+    var<workgroup> KV_j : array<array<${qInput.type.storage}, ${workgroupSize[0]}>, ${bC}>;
+    var<workgroup> S_i_j : array<array<${output.type.storage}, ${bC}>, ${bR}>;
+  ${shaderHelper.registerUniforms(uniforms).declareVariables(...inputVars, ...outputVars)}
+  ${shaderHelper.mainStart(workgroupSize)}
+    let offset = (workgroup_id.z * uniforms.batchOffset + workgroup_id.y * uniforms.headOffset) / ${components} + workgroup_id.x * ${bR} * uniforms.d;
+
+    var O_i : array<${type}, ${numTiles * colsPerThread}>;
+    var l_i_j = ${type}(0);
+    var m_i_j = ${type === 'f32' ? 'f32(-3.402823e+38f)' : 'f16(-65504)'};
+
+    for (var tile = 0; tile < ${numTiles}; tile++) {
+      Q_i[local_id.y][u32(${workgroupSize[0]} * tile) + local_id.x] = Q[offset + local_id.y * uniforms.d + u32(${workgroupSize[0]} * tile) + local_id.x];
+    }
+
+    for (var j = 0; j < ${tC}; j++) {
```

Comment:
Turn tC to uniform or add it to hint?

Reply:
Thanks. I also noticed this issue. I added it to uniform.